### ISSUE TITLE
split Guard class into two

### ISSUE THREAD

**Issue by romani** [checkstyle/checkstyle#3497]
https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/Guard.java

it serves two differend modes that are managed by null values in fields.

Task: create two classes from one to avoid checking for null. Names should be different "Guard" and named close to what they represent from XML.

**Comment by jochenvdv** [comment]
Hi, I would like to tackle this issue as my first contribution to the checkstyle project as this seems relatively simple to begin with.

As I see it, there should be at least two different classes: a `ClassGuard` and a `PkgGuard`.
Should they both inherit from an abstract `Guard` class? The `calculateResult` method could be used by both.
The unit test would also have to be split in two.

If there is any other information I would need to implement this, please let me know.

**Comment by romani** [comment]
Thanks for desire to help, you always welcome. Do not hesitate to ask questions.

>Should they both inherit from an abstract Guard class? 

Please propose better names. "Guard" is not good. It is better to less in common, overlap in one field is not enough to make an abstract class for them.

>The unit test would also have to be split in two.

Yes

> If there is any other information I would need to implement this, please let me know.


Please do PR , and we will see what to change. You can skip uts for first review

**Comment by jochenvdv** [comment]
> Please propose better names. "Guard" is not good.

What do you think of the names `ClassImportRule` and `PkgImportRule`? Or maybe `ClassImportAccessRule` and `PkgImportAccessRule`?


> It is better to less in common, overlap in one field is not enough to make an abstract class for them.

Ok, but unless we create and abstract class or an interface we cannot operate on the general concept of a 'Guard' anymore from the `PkgControl` class.
Right now all Guard objects are stored in a single `Deque` (LinkedList), do you want them to be stored and handled separately?. We could also store both of them in a `Deque(Object)` and then use instanceof and typecasting to work with them.

**Comment by jochenvdv** [comment]
I have started working on this.

- `GuardTest.java` contained some tests for `AccessResult`, I move them into its own test file
- I think `PkgControl` should also be renamed to `ImportControl`
- The unit tests for `PkgControl` only test for packages, so I will also add some tests for classes

I'm gonna do what I think is best for now. I will make a PR once references to `Guard` and `PkgControl` have been replaced, I can then still change things if needed.
